### PR TITLE
Update user delete response

### PR DESCRIPTION
**Pull request by peiyaoh**
### Description

- Update user delete return value to be boolean to align with data manager.

### Related Jira Tickets

- https://umich-mia.atlassian.net/browse/JST-97?atlOrigin=eyJpIjoiMThkMzhjNzBjNWQ3NGYwNzkyYWExYmZkODA5NjgwZDciLCJwIjoiaiJ9

### Comments 

-

Please ensure all items are complete before requesting a review:

- [ ] **Code Completed**: All code for the feature or bugfix has been implemented.
- [ ] **Unit Tests Added**: Unit tests are written for all new functionality, covering edge cases where applicable.
- [ ] **Integration Tests (where necessary)**: Integration tests are added for interactions between modules or components.
- [ ] **All Tests Passing**: All unit and integration tests are passing.
- [ ] **Type Safety Checked**: Ensure TypeScript types are correctly defined and used. Avoid `any` types without proper justification.
- [ ] **Typedoc Comments Added**: Ensure all functions, classes, interfaces, and types are documented using TypeDoc syntax.
- [ ] **Code Formatting**: All code files are formatted according to the project’s standards (e.g., Prettier, ESLint) and no linting errors remain.
- [ ] **No Console Logs**: Remove any `console.log` or other debugging output from the code.
- [ ] **Documentation Updated**: Update relevant project documentation (e.g., README, configuration guides) if necessary.

---

### CODE REVIEW

File: src/user-manager/user-manager.ts [hunk 1]
```diff
@@ -316,10 +316,11 @@ const updateUserById = async (
  * @param {string} userId - The user's ID.
  * @returns {Promise<void>} Resolves when deletion is complete.
  */
-const deleteUserById = async (userId: string): Promise<void> => {
+const deleteUserById = async (userId: string): Promise<boolean> => {
   _checkInitialization();
-  await dm.removeItemFromCollection(USERS, userId);
+  const result = await dm.removeItemFromCollection(USERS, userId);
   _users.delete(userId);
```
Comment (jadrubio) [inline src/user-manager/user-manager.ts]: We should do a check to see if the result is true before removing it from _users

File: src/user-manager/user-manager.ts [hunk 2]
```diff
@@ -329,10 +330,11 @@ const deleteUserById = async (userId: string): Promise<void> => {
  * @param {string} userId - The user's ID.
  * @returns {Promise<void>} Resolves when deletion is complete.
  */
-const deleteUserByUniqueIdentifier = async (uniqueIdentifier: string): Promise<void> => {
+const deleteUserByUniqueIdentifier = async (uniqueIdentifier: string): Promise<boolean> => {
   const theUser: JUser = await getUserByUniqueIdentifier(uniqueIdentifier) as JUser;
-  await deleteUserById(theUser.id);
+  const result = await deleteUserById(theUser.id);
   _users.delete(theUser.id);
```
Comment (jadrubio) [inline src/user-manager/user-manager.ts]: We should do a check to see if the result is true before removing it from _users

Comment (peiyaoh) [reply src/user-manager/user-manager.ts]: You are right. Let me add it.

File: src/user-manager/user-manager.ts (hunk 1 above)
Comment (peiyaoh) [reply src/user-manager/user-manager.ts]: You are right. Let me add it.

Comment (peiyaoh) [reply src/user-manager/user-manager.ts]: @jadrubio Just add the check. Please take a look.

One interesting aspect of defining behavior (in MongoDB) is that the `deleteOne` method will return `acknowledged == true` even if no item matched the deletion criteria. In a sense, it behaves as if the item was deleted. I am trying to replicate this behavior in the UserManager as well.

File: src/user-manager/user-manager.ts (hunk 2 above)
Comment (peiyaoh) [reply src/user-manager/user-manager.ts]: @jadrubio Just add the check. Please take a look.

**Review by jadrubio — DISMISSED** [review]

File: src/user-manager/user-manager.ts [hunk 3]
```diff
@@ -314,25 +314,28 @@ const updateUserById = async (
...
  */
-const deleteUserByUniqueIdentifier = async (uniqueIdentifier: string): Promise<void> => {
-  const theUser: JUser = await getUserByUniqueIdentifier(uniqueIdentifier) as JUser;
-  await deleteUserById(theUser.id);
-  _users.delete(theUser.id);
+const deleteUserByUniqueIdentifier = async (uniqueIdentifier: string): Promise<boolean> => {
+  const theUser: JUser | null = await getUserByUniqueIdentifier(uniqueIdentifier);
+  const userId = theUser?.id as any;
+  const result = await deleteUserById(userId);
+  if(result) _users.delete(userId);
```
Comment (numerator) [inline src/user-manager/user-manager.ts:337]: Doesn't `deleteUserById()` already delete from the cache?

File: src/user-manager/user-manager.ts [hunk 4]
```diff
@@ -314,25 +314,28 @@ const updateUserById = async (
...
  * @param {string} userId - The user's ID.
- * @returns {Promise<void>} Resolves when deletion is complete.
+ * @returns {Promise<boolean>} Resolves to true if deletion was successful, false otherwise.
  */
-const deleteUserById = async (userId: string): Promise<void> => {
+const deleteUserById = async (userId: string): Promise<boolean> => {
   _checkInitialization();
-  await dm.removeItemFromCollection(USERS, userId);
-  _users.delete(userId);
+  const result = await dm.removeItemFromCollection(USERS, userId);
```
Comment (numerator) [inline src/user-manager/user-manager.ts:321]: digging into this, I see that this is the code for `removeItemFromCollection` in mongo-data-manager:
```
const removeItemFromCollection = async (
  collectionName: string,
  id: string
): Promise<boolean> => {
  ensureInitialized();
  const objectId = toObjectId(id);
  if (!objectId) return false;

  try {
    const { acknowledged } = await _db!
      .collection(collectionName)
      .deleteOne({ _id: objectId });
    return acknowledged;
  } catch (error) {
    return handleDbError(
      `Error removing item with id ${id} from ${collectionName}`,
      error
    );
  }
};
```

Are we sure that "acknowledged" will be false if the doc is not actually removed? Like if an objectId is provided but that document doesn't exist? I think the result returned from `deleteOne()` in such a case will be `{acknowledged: true, deletedCount: 0}`. It would be good to verify how this works and, if I'm right, what should we do?

File: src/user-manager/user-manager.ts (hunk 3 above)
Comment (peiyaoh) [reply src/user-manager/user-manager.ts:337]: Good point. I just removed cache deletion from this function.

File: src/user-manager/user-manager.ts (hunk 4 above)
Comment (peiyaoh) [reply src/user-manager/user-manager.ts:321]: Based on my read of the MongoDB documentation, acknowledged == true means the delete option has been attempted. The deletedCount represents the actual number of documents deleted (provided that "acknowledged" is true).

I think we can stay with the current definition, and add a warning when deletedCount == 0 (document not found, effectively complete the delete) so that devs are aware.

This follows the same model in updateItem, when the update operation is attempted that does not change any values (effectively complete the update), we log the warning and still return the document.

Alternative 1: `return deletedCount > 0;` to indicate whether an item has been deleted.

Alternative 2: we could still return the document (if found) || null (not found), so that devs can have more information. Mongo provides findOneAndDelete() for this. It does require a write lock, which could impact performance.
https://www.mongodb.com/docs/manual/reference/method/db.collection.findOneAndDelete/


Terminology

Acknowledged: A boolean acknowledged as true if the operation ran with write concern or false if write concern was disabled.
https://www.mongodb.com/docs/manual/reference/method/db.collection.deleteOne/

Write concern
Specifies whether a write operation has succeeded. Write concern allows your application to detect insertion errors or unavailable MongoDB instances. 
https://www.mongodb.com/docs/manual/reference/glossary/#std-term-write-concern

Write concern describes the level of acknowledgment requested from MongoDB for write operations to a standalone mongod, replica sets, or sharded clusters.
https://www.mongodb.com/docs/manual/reference/write-concern/

**Comment by peiyaoh** [comment]
Hi all,

My digging revealed a nuance in our current implementation regarding the mongo-data-manager deletion logic. I want to highlight the current behavior of this PR.

🔍 Current Deletion Behavior
The mongo-data-manager has two different outcomes for unsuccessful deletions:
- Invalid ID: An invalid ID throws an error at the manager level and returns false. (This signals a technical failure).
- No Match Found: If the ID is valid but no item exists, it returns true (Success - the item is gone).

PR Behavior: This PR, by propagating the result up to user-manager.deleteUserByUniqueIdentifier(), maintains this logic.

⚠️ Impact of Current Logic (Double Deletion)
If a deletion of the same existing user is attempted twice:
- The first attempt will return true (Successful deletion).
- The second attempt will return false (Unsuccessful—no matching item found, but due to the inherited logic path from the called method as deleteUserByUniqueIdentifier can no longer find a matching user with a valid id).

Good News: In this scenario, the second attempt logs an error but does not crash.

💡 Next Steps
I suggest we leave any consideration of changing this behavior until after the core + engine refactoring is complete.

Let me know if you have any questions or concerns about merging this PR with its current behavior.
